packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts

```diff
@@ -199,28 +228,51 @@ export const buildAggregatedFilterItemsApplier = (
     return null;
   }
 
-  return (rowId, shouldApplyFilter) => {
-    const resultPerItemId: GridFilterItemResult = {};
+  if (appliers.length === 1) {
+    const applier = appliers[0];
+    const applierFn = applier.fn;
 
-    const filteredAppliers = shouldApplyFilter
-      ? appliers.filter((applier) => shouldApplyFilter(applier.item.field))
-      : appliers;
+    const applierCall = applier.v7
+      ? 'applierFn(row)'
+      : 'applierFn(getRowId ? getRowId(row) : row.id)';
+    const fn = eval(`
+      (row, shouldApplyFilter) => {
+        // ${applierFn.name} <- Keep a ref, prevent the bundler from optimizing away
 
-    filteredAppliers.forEach((applier) => {
-      resultPerItemId[applier.item.id!] = applier.fn(rowId);
-    });
+        if (shouldApplyFilter && !shouldApplyFilter(applier.item.field)) {
+          return { '${applier.item.id!}': false };
+        }
+        return { '${applier.item.id!}': ${applierCall} };
+      }
+    `);
+
+    return fn;
```

Comment:
So if we consider the final results as the baseline, and we decide to not apply this change, the performance degrades by 20%. I know "eval is evil", but in this case we're getting a real performance improvement, so I feel like we should keep it.

Reply:
Exotic indeed, but also risky - try running this locally 😅

```tsx
import * as React from 'react';
import { DataGrid } from '@mui/x-data-grid';

const columns = [{ field: 'id' }];

export default function InitialFilters() {
  const [rows, setRows] = React.useState<any[]>([]);

  React.useEffect(() => {
    setRows([{ id: 1 }]);
  }, []);

  return (
    <div style={{ height: 400, width: '100%' }}>
      <DataGrid
        columns={columns}
        rows={rows}
        filterModel={{
          items: [
            {
              field: 'id',
              operator: 'equals',
              value: '1',
              id: "1': alert('hello') } //",
            },
          ],
        }}
      />
    </div>
  );
}
```

Reply:
Fair point! Could we reliably escape that value with `... { ${JSON.stringify(String(applier.item.id))}: ...`? I feel that if `String` is guaranteed to return a string, then `JSON.stringify` should be guaranteed to return valid javascript string representation.

Reply:
I'm not sure I follow.
`${JSON.stringify(String(applier.item.id))}` would return the same string, and it will be evaluated as before. Am I missing something?

Reply:
`JSON.stringify` is doing the string-quoting, which means it's escaped properly for javascript:

![image](https://github.com/mui/mui-x/assets/1423607/54f9be22-f434-4b8b-9ec6-f33b63e30d70)

Alternatively using `parseInt(applier.item.id)` could also work. It would return `NaN` for invalid payloads, and it's guaranteed to return a `number`.